src/benchmarklib/benchmark_runner.cpp

```diff
@@ -0,0 +1,449 @@
+  }
+
+  auto mutable_named_queries = _queries;
+
+  // For shuffling the query order
+  std::random_device random_device;
+  std::mt19937 random_generator(random_device());
+
+  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
+  while (state.keep_running()) {
+    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
+
+    for (const auto& named_query : mutable_named_queries) {
+      const auto query_benchmark_begin = std::chrono::steady_clock::now();
+
+      // Execute the query, we don't care about the results
+      _execute_query(named_query);
+
+      const auto query_benchmark_end = std::chrono::steady_clock::now();
+
+      auto& query_benchmark_result = _query_results_by_query_name.at(named_query.first);
+      query_benchmark_result.duration += query_benchmark_end - query_benchmark_begin;
+      query_benchmark_result.num_iterations++;
+    }
+  }
+}
+
+void BenchmarkRunner::_benchmark_individual_queries() {
+  for (const auto& named_query : _queries) {
+    const auto& name = named_query.first;
+    _config.out << "- Benchmarking Query " << name << std::endl;
+
+    BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
+    while (state.keep_running()) {
+      _execute_query(named_query);
+    }
+
+    QueryBenchmarkResult result;
+    result.num_iterations = state.num_iterations;
+    result.duration = std::chrono::high_resolution_clock::now() - state.begin;
+
+    _query_results_by_query_name.emplace(name, result);
+  }
+}
+void BenchmarkRunner::_execute_query(const NamedQuery& named_query) {
+  const auto& name = named_query.first;
+  const auto& sql = named_query.second;
+
+  auto pipeline = SQLPipelineBuilder{sql}.with_mvcc(_config.use_mvcc).create_pipeline();
+  // Execute the query, we don't care about the results
+  pipeline.get_result_table();
+
+  // If necessary, keep plans for visualization
+  if (_config.enable_visualization) {
+    const auto query_plans_iter = _query_plans.find(name);
+    if (query_plans_iter == _query_plans.end()) {
+      Assert(pipeline.get_query_plans().size() == 1, "Expected exactly one SQLQueryPlan");
+      QueryPlans plans{pipeline.get_optimized_logical_plans(), pipeline.get_query_plans()};
+      _query_plans.emplace(name, plans);
+    }
+  }
+}
+
+void BenchmarkRunner::_create_report(std::ostream& stream) const {
+  nlohmann::json benchmarks;
+
+  for (const auto& named_query : _queries) {
+    const auto& name = named_query.first;
+    const auto& query_result = _query_results_by_query_name.at(name);
+
+    const auto duration_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(query_result.duration).count();
+    const auto duration_seconds = static_cast<float>(duration_ns) / 1'000'000'000;
+    const auto items_per_second = static_cast<float>(query_result.num_iterations) / duration_seconds;
+    const auto time_per_query = duration_ns / query_result.num_iterations;
+
+    nlohmann::json benchmark{
```

Comment:
Can you also include the settings (command line string is fine) and the current GIT hash?

Reply:
- the settings are included in the context. 
- how do you suggest i get the git hash?

Reply:
i added all missing settings to the context

Reply:
Can't we get cmake to add `-D GIT_HASH=blabla12`?

Reply:
That would get a bit hacky, cmake only re-runs itself when one of its own files changed. You can add a pseudo target that gets triggered when a specific file in .git changes - you'd usually make it generate a header with the GIT tag which you then include in the project.

Reply:
https://cmake.org/cmake/help/v3.11/command/add_custom_command.html

Reply:
How hacky is that? I see a value in having the commit id so that when we find old benchmark results somewhere, we know which version they used.

Reply:
I something similar for creating a latest-change timestamp from all source files and it was BAD. Git versioning should be simpler - I searched a bit and this guy (https://github.com/andrew-hardin/cmake-git-version-tracking) seems to want to do the right things 
* git tags are stored in a normal header/source file
* no recompilation of anything if git state didn't change
* no dependency on manually triggering cmake

I haven't read all his source (https://github.com/andrew-hardin/cmake-git-version-tracking/blob/master/git_watcher.cmake) but it looks roughly like what I would expect to see.

Reply:
Whereas on the cmake mailing list they calmly - and with no one weighing in sanity - propose approaches that will only generate the version header on configuration time https://cmake.org/pipermail/cmake/2010-July/037958.html

Reply:
I kinda can't tell which solution is be preferred? what do you want/need me to do here? this all seems hacky to me

Reply:
If @mrks wants git-rev tagging of benchmark results, go for andrew-hardins approach, probably c/p it into cmake/

Reply:
out context now looks like this:
```
"context": {
    "GIT-HASH": "6c6b5b7db5e0996812e17208ad11b37894850b21",
    "benchmark_mode": "IndividualQueries",
    "build_type": "debug",
    "chunk_size": 4294967294,
    "date": "2018-05-24 15:51:10",
    "encoding": "dictionary",
    "max_duration (s)": 5,
    "max_runs": 1000,
    "output_file_path": "stdout",
    "scale_factor": 0.00100000004749745,
    "using_mvcc": false,
    "using_scheduler": false,
    "using_visualization": false,
    "verbose": false
  }
```

I have no idea how or why this cmake git magic works, but it works on my machine.

Reply:
Does it get a "-dirty" tag if the working dir has local changes? I think that would be pretty important to avoid confusion

Reply:
Can be done. Will be done. Just need to add it to the git hash string

Reply:
Sweet. Also linking #63 so that when we get around to look at that, we include all relevant knobs here.